src/main/java/mat/server/service/impl/FhirCqlLibraryServiceImpl.java

```diff
@@ -0,0 +1,158 @@
+        this.cqlVisitorFactory = cqlVisitorFactory;
+    }
+
+    @Override
+    public FhirConvertResultResponse convertCqlLibrary(CQLLibraryDataSetObject sourceLibrary, String loggedInUser) throws MatException, MarshalException, MappingException, IOException, ValidationException {
+
+        if (!sourceLibrary.isFhirConvertible()) {
+            throw new MatException("Cql Library cannot be converted to FHIR");
+        }
+
+        fhirConvertResultResponse.setSourceLibrary(sourceLibrary);
+        cqlLibraryServiceInterface.isLibraryAvailableAndLogRecentActivity(sourceLibrary.getId(), loggedInUser);
+
+        CQLLibrary existingLibrary = cqlLibraryDAO.find(sourceLibrary.getId());
+
+        /*if draft FHIR library already exists then delete it */
+        deleteDraftFhirLibrariesInSet(existingLibrary.getSetId());
+
+        ConversionResultDto conversionResult = fhirLibraryConversionRemoteCall.convert(sourceLibrary.getId(), ConversionType.CONVERSION);
+        Optional<String> fhirCqlOpt = getFhirCql(conversionResult);
+
+        FhirValidationStatus validationStatus = createValidationStatus(conversionResult);
+        fhirConvertResultResponse.setValidationStatus(validationStatus);
+
+        if (!fhirCqlOpt.isPresent()) {
+            // If there is no FHIR CQL, then we don't persist the Library. FHIR Library cannot be created.
+            throw new MatException("Your library cannot be converted to FHIR. Outcome: " + validationStatus.getOutcome() + " Error Reason: " + validationStatus.getErrorReason());
+        } else {
+            /*Create a new cql Library object and store it in DB*/
+            persistFhirLibrary(loggedInUser, existingLibrary, fhirCqlOpt);
+        }
+        return fhirConvertResultResponse;
+    }
+
+    private FhirValidationStatus createValidationStatus(ConversionResultDto convertResult) {
+        FhirValidationStatus validationSummary = new FhirValidationStatus();
+        validationSummary.setErrorReason(convertResult.getErrorReason());
+        validationSummary.setOutcome(convertResult.getOutcome() != null ? convertResult.getOutcome().toString() : null);
+        validationSummary.setValidationPassed(ConversionOutcome.SUCCESS.equals(convertResult.getOutcome()));
+        return validationSummary;
+    }
+
+    private Optional<String> getFhirCql(ConversionResultDto conversionResult) {
+        return Optional.ofNullable(conversionResult.getLibraryConversionResults()).stream()
+                .flatMap(libConvRes -> libConvRes.stream())
+                .map(cqlLibRes -> cqlLibRes.getCqlConversionResult())
+                .filter(el -> el != null)
+                .map(el -> el.getFhirCql())
+                .filter(StringUtils::isNotBlank)
+                .findFirst();
+    }
+
+    private void persistFhirLibrary(String loggedInUser, CQLLibrary existingLibrary, Optional<String> fhirCqlOpt) throws MatException, MarshalException, MappingException, ValidationException, IOException {
```

Comment:
I recall there were some issues with 'persistFhirMeasure'. Its steps were not executed within the same transaction and the change was not atomic. An error in the middle could cause artifacts  - like a persisted FHIR measure with QDM CQL ..
So, you might want to double-check if all the steps are executed within the same transaction in  persistFhirLibrary.

Reply:
Maybe not needed in that case.